### PR TITLE
Dependency injection for mocking getVariable and setVariable

### PR DESCRIPTION
**Pull request by hashtagchris**
Add a `KeyValueStoreInterface` interface and use different implementations for task execution and mocking.

### CODE REVIEW

**Comment by hashtagchris** [comment]
Putting an in-progress draft out to get early feedback on my approach. I'm adding `KeyValueStoreInterface` which will have two implementations - one that uses `process.env` as a backing store, and another that uses `answers.variables`. (See #521.) I think the same interface could be used for secrets as well.

To inject the right implementation, I've come up with two solutions:

A. Add both dependencies.ts and mock-dependencies.ts. Have mock-run register an instance of mock-dependencies.ts in addition to mock-task.ts. The rest of the code can require/import `./dependencies`. I have the start of that solution in this PR.

B. Refactor the code currently in task.ts into a class, and have the constructor require dependencies. Have task.ts create an instance of that class and export it. Example:

task.ts
```typescript
import { TaskImpl } from "./TaskImpl";
import { ProcessKeyValueStore } from "./ProcessKeyValueStore";

const task = new TaskImpl(new ProcessKeyValueStore());

// export an instance of this class as a collection of functions.
export = task;
```

mock-task.ts - perhaps this can be simplified?
```typescript
import { TaskImpl } from "./TaskImpl";

export function setAnswers(answers: ma.TaskLibAnswers) {
    let task: TaskImpl;
    if (answers.variables) {
        task = new TaskImpl(new MockKeyValueStore(answers.variables));
    }
    else {
        // If no variable answers were supplied, continue to use `process.env` to 
        // avoid breaking existing unit tests.
        task = new TaskImpl(new ProcessKeyValueStore());
    }

    return {
        // mock implementations of functions
        exist: function (path: string): boolean {
            return mock.getResponse('exist', path, module.exports.debug) || false;
        },
        ...

        // proxy some functions to task.
        setStdStream: task.setStdStream.bind(task),
        setErrStream: task.setErrStream.bind(task),
        ...
    }
}
```

Solution A is the quicker solution. Less changes, less risk, easier to port.

Solution B seems like a more pure form of dependency injection, and the code might be more predictable. But it will involve moving a lot of code around, making it harder to port to the releases/2.8.0 branch. Also mock-task.ts has a mix of its own functions and proxies to task implementations, so it won't be as clean as task.ts with a bigger redesign.

Thoughts? Am I missing a solution?

**Comment by damccorm** [comment]
I like solution A a lot here, in particular because we've decided not to major version at this point (we haven't actually introduced breaking changes yet), so low churn is good. Overall this feels clean though, thanks for getting this rolling.

**Comment by hashtagchris** [comment]
Thanks. Any tips on testing task-lib changes with real tasks?

**Comment by hashtagchris** [comment]
Thanks. I'm starting to warm on solution B, after hearing it likely won't have to be ported and realizing shortly before I posted how a subclass makes it easier to fall through to task.ts's implementations. But I'm open to either.

@bryanmacfarlane @brcrista Any feedback before I continue work?

File: node/task.ts [hunk 1]
```diff
@@ -50,6 +53,8 @@ export enum Platform {
     Linux
 }
 
+const _variableStore: KeyValueStoreInterface = dependencies.variableStore;
```
Comment (brcrista) [inline node/task.ts]: I don't think this is quite right -- clearly this module has a bunch of depenencies that don't come through `dependencies`.

You can use the same mockery pattern here as with everything else.

```
// variablestore.ts
export variableStore = process;

// task.ts
import { variableStore } from './variablestore';

// reference `variableStore` in the code
```

**Comment by brcrista** [comment]
I noticed that this only mocks out one line of `getVariable` and `setVariable`.  Is there anything preventing us from replacing the entire functions with mock-task implementations?  I feel like they should just read and write to an in-memory store that can be initialized with some key-value pairs.

File: node/interfaces.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+// Generic key-value store. May or may not be case-sensitive.b
+export interface KeyValueStoreInterface {
```
[the remark this answers is not in the document]
Comment (hashtagchris) [reply node/interfaces.ts]: I was following an example at http://www.typescriptlang.org/docs/handbook/interfaces.html too closely, but I think the leading `I` is a C#ism. I'll rename to `KeyValueStore`.

**Comment by hashtagchris** [comment]
There's a lot of logic in `getVariable`, `setVariable` and `setSecret` which I started to duplicate in #521 - normalizing the key, checking if the variable is marked secret, checking for multi-line secrets, etc. I think ideally all that logic should live in one layer and then call into variable/secret store for simple read-write-delete operations. That makes it easy to keep the mocks high fidelity.

File: node/interfaces.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (brcrista) [reply node/interfaces.ts]: Oh yeah, you're right.  I had thought TS followed that convention as well but it doesn't.

**Comment by brcrista** [comment]
I see.  I was just wondering if stuff like the vault is really testable.

**Comment by stephenmichaelf** [comment]
Are we still planning to merge this?

**Comment by stephenmichaelf** [comment]
Closing due to inactivity.

**Pull request closed without merging**